File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        int len = board[0].length;
         Stack<Integer> st = new Stack<>();
         
        for(int mv : moves){
            mv -= 1;
            for(int i=0; i < len; i++){  
                if(board[i][mv] != 0){ //인형이 있는경우
                    if(st.size() > 0 && st.peek() == board[i][mv] ){//뽑은 인형 == 마지막 인형
                        st.pop();
                        answer += 2;
                    }else st.push(board[i][mv]); //인형 넣기 
                    
                    board[i][mv] = 0; //뽑은 뒤 0
                    break; 
                }
            }  
        } 
        
        return answer;
    }
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.js
function solution(board, moves) {
    let answer = 0;
    let stack = [];

    for (let move of moves) {
        let mv = move - 1;
        
        for (let i = 0; i < board.length; i++) {
            let doll = board[i][mv];
            
            if (doll !== 0) {
                if (stack[stack.length - 1] === doll) {
                    stack.pop();
                    answer += 2;
                } else {
                    stack.push(doll);
                }

                board[i][mv] = 0;
                break;
            }
        }
    }

    return answer;
}
